optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java

```diff
@@ -69,6 +78,14 @@ public void addAllDestinations(Object origin, Iterator<Object> destinationIterat
 
     public Object getDestination(Object origin, int nearbyIndex) {
         Object[] destinations = originToDestinationsMap.get(origin);
+        if (destinations == null) {
+            /*
+             * The item may be missing in the distance matrix due to an underlying filtering selector.
+             * In such a case, the distance matrix needs to be updated.
+             */
+            addAllDestinations(origin);
+            destinations = originToDestinationsMap.get(origin);
```

Comment:
If it were me, I'd just recurse at this point - calling the same function on the new data, making sure it works.

Reply:
+0 if, for any reason, the matrix is not updated, the current approach ends up with NPE, the approach you're suggesting ends up with StackOverflowError.

Reply:
Fair enough.